File: esim-frontend/esim-ide/schematic/elementpin.cc
/******************************************************************************
 * OpenESIM (Electrical Engineering Simulator)                                *
 * Copyright (C) 2019, ESIM team. <diyer175@hotmail.com>                      *
 *                                                                            *
 * This project is free software; you can redistribute it and/or              *
 * modify it under the terms of the GNU Lesser General Public License(GPL)    *
 * as published by the Free Software Foundation; either version 2.1           *
 * of the License, or (at your option) any later version.                     *
 *                                                                            *
 * This project is distributed in the hope that it will be useful,            *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of             *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU          *
 * Lesser General Public License for more details.                            *
 ******************************************************************************/

#include "elementpin.h"
#include "renderdevice.h"
#include "persistencedom.h"
#include "errtrace.h"
#include "util.h"
#include "elementtext.h"
#include "schematic.h"
#include <cmath>

namespace schematic
{

ElementPin::ElementPin(int x1, int y1) :
    Element(),
    m_x1(x1),
    m_y1(y1),
    m_length(10 * 2),
    m_symbolSpacing(5),
    m_numSpacing(2),
    m_symbol("PIN"),
    m_num("1"),
    m_angle(0),
    m_elecType(util::ELEC_INPUT),
    m_pinType(util::PIN_NONE),
    m_lineX1(0),
    m_lineY1(0),
    m_lineX2(0),
    m_lineY2(0),
    m_lineStyle(*(StyleTemplates::getInstance()->getLineStyleRef("pin"))),
    m_pinSymbolText(new ElementText),
    m_pinNumText(new ElementText)
{
  m_pinSymbolText->setText(m_symbol);
  m_pinSymbolText->setTextStyle(*(StyleTemplates::getInstance()->getTextStyleRef("symbol")));

  m_pinNumText->setText(m_num);
  m_pinNumText->setTextStyle(*(StyleTemplates::getInstance()->getTextStyleRef("pin")));

}

const char *ElementPin::classname() const
{
  return "pin";
}

int ElementPin::locationType() const
{
  return S0_POS_1P | S1_ROTATION_1A;
}

void ElementPin::locate(int opcode, int paramSize, ...)
{
  va_list vlist;
  va_start(vlist, paramSize);
  switch(opcode)
  {
    case locateSetPos:
      ASSERT_PARAMS_RETURN (paramSize == 2 * sizeof(int));

      m_x1 = va_arg(vlist, int);
      m_y1 = va_arg(vlist, int);
      relayout();
      break;
    case locateGetPos:
      ASSERT_PARAMS_RETURN (paramSize == 2 * sizeof(int*));

      *va_arg(vlist, int*) = m_x1;
      *va_arg(vlist, int*) = m_y1;
      break;
    case locateSetRotation:
      ASSERT_PARAMS_RETURN (paramSize == 1 * sizeof(int));
        {
          int ca = va_arg(vlist, int);
          int angle = std::abs(ca);
          int coeff = ca > 0 ? 1 : -1;
          while (angle > 360)
            { angle -= 360; }
          if (0 <= angle && angle < 90) // only do angles in four directios are used
            { angle = 0; }
          else if (90 <= angle && angle < 180)
            { angle = 90; }
          else if (180 <= angle && angle < 270)
            { angle = 180; }
          else if (270 <= angle && angle < 360)
            { angle = 270; }
          angle *= coeff;
          m_angle = (angle < 0) ? (angle + 360) : angle; // normalize the angle to positive
          relayout();
        }
      break;
    case locateGetRotation:
      ASSERT_PARAMS_RETURN (paramSize == 1 * sizeof(int*));

      *va_arg(vlist, int*) = m_angle;
      break;
  }
  va_end(vlist);
  return;
}

/**
 * @brief Layout the pin paintings in four directions
    called when the layout has been changed.
 */
void ElementPin::relayout()
{
  m_pinSymbolText->locate(locateSetRotation, sizeof(int), m_angle);
  m_pinNumText->locate(locateSetRotation, sizeof(int), m_angle);

  m_lineX1 = m_x1; m_lineY1 = m_y1;

  int cx, cy;

  /*
   * Layout:
   *               NumText(center)
   * [ ]  ( ) -----------------------   SymbolText
   *  |    |
   * note port symbol
   */
  switch (m_angle)
  {
    case 180: // left
      m_lineX2 = m_x1 - m_length;
      m_lineY2 = m_y1;

      cx = m_lineX1 - (m_length - m_pinNumText->getTextWidth())/2;
      cy = m_lineY1 - m_pinNumText->getTextHeight() - m_numSpacing;
      m_pinNumText->locate(locateSetPos, 2 * sizeof(int), cx ,cy);

      cx = m_lineX2 - m_symbolSpacing;
      cy = m_lineY2 - (m_pinNumText->getTextHeight()/2);
      m_pinSymbolText->locate(locateSetPos, 2 * sizeof(int), cx ,cy);
      break;

    case 90: // top
      m_lineX2 = m_x1;
      m_lineY2 = m_y1 + m_length;

      cy = m_lineY1 + (m_length - m_pinNumText->getTextWidth())/2;
      cx = m_lineX1 - m_numSpacing;
      m_pinNumText->locate(locateSetPos, 2 * sizeof(int), cx, cy);

      cy = m_lineY1 + m_length + m_symbolSpacing;
      cx = m_lineX1 + (m_pinNumText->getTextHeight()/2);
      m_pinSymbolText->locate(locateSetPos, 2 * sizeof(int), cx, cy);
      break;

    case 0: // right
      m_lineX2 = m_x1 + m_length;
      m_lineY2 = m_y1;

      cx = m_lineX1 + (m_length - m_pinNumText->getTextWidth())/2;
      cy = m_lineY1 - m_pinNumText->getTextHeight() - m_numSpacing;
      m_pinNumText->locate(locateSetPos, 2 * sizeof(int), cx, cy);

      cx = m_lineX2 + m_symbolSpacing;
      cy = m_lineY2 - (m_pinNumText->getTextHeight()/2);
      m_pinSymbolText->locate(locateSetPos, 2 * sizeof(int), cx, cy);
      break;

    case 270:
      m_lineX2 = m_x1;
      m_lineY2 = m_y1 - m_length;

      cy = m_lineY2 + (m_length+m_pinNumText->getTextWidth())/2;
      cx = m_lineX1 - m_pinNumText->getTextHeight() - m_numSpacing;
      m_pinNumText->locate(locateSetPos, 2 * sizeof(int), cx, cy);

      cy = m_lineY2 - m_symbolSpacing;
      cx = m_lineX2 - (m_pinNumText->getTextHeight()/2);
      m_pinSymbolText->locate(locateSetPos, 2 * sizeof(int), cx, cy);
  }
}

void ElementPin::render(RenderDevice *device)
{
  LineStyle oldline = device->setLineStyle(m_lineStyle);

  if (this->isModifiable())
    { device->ellipse(m_lineX1 - 5, m_lineY1 - 5, 10, 10); }
  device->line(m_lineX1, m_lineY1, m_lineX2, m_lineY2);

  device->setLineStyle(oldline);
}

void ElementPin::bounding(int *x1, int *y1, int *x2, int *y2) const
{
  *x1 = m_lineX1; *y1 = m_lineY1;
  *x2 = m_lineX2, *y2 = m_lineY2;

  *x1 = s_min(*x1, *x2); *y1 = s_min(*y1, *y2);
  *x2 = s_max(*x1, *x2), *y2 = s_max(*y1, *y2);
}

int ElementPin::serialize(persistence::PersistenceDom *dom)
{
  RC_CHECK_RETURN(Element::serialize(dom));

  RC_CHECK_RETURN(dom->ser(m_x1));
  RC_CHECK_RETURN(dom->ser(m_y1));
  RC_CHECK_RETURN(dom->ser(m_length));
  RC_CHECK_RETURN(dom->ser(m_symbolSpacing));
  RC_CHECK_RETURN(dom->ser(m_numSpacing));
  RC_CHECK_RETURN(dom->ser(m_symbol));
  RC_CHECK_RETURN(dom->ser(m_num));
  RC_CHECK_RETURN(dom->ser(m_angle));
  RC_CHECK_RETURN(dom->ser((int)m_elecType));
  RC_CHECK_RETURN(dom->ser((int)m_pinType));

  RC_CHECK_RETURN(m_lineStyle.serialize(dom));

  RC_CHECK_RETURN(m_pinNumText->serialize(dom));
  RC_CHECK_RETURN(m_pinSymbolText->serialize(dom));
  return 0;
}

int ElementPin::deserialize(persistence::PersistenceDom *dom)
{
  RC_CHECK_RETURN(Element::deserialize(dom));

  RC_CHECK_RETURN(dom->des(&m_x1));
  RC_CHECK_RETURN(dom->des(&m_y1));
  RC_CHECK_RETURN(dom->des(&m_length));
  RC_CHECK_RETURN(dom->des(&m_symbolSpacing));
  RC_CHECK_RETURN(dom->des(&m_numSpacing));
  RC_CHECK_RETURN(dom->des(&m_symbol));
  RC_CHECK_RETURN(dom->des(&m_num));
  RC_CHECK_RETURN(dom->des(&m_angle));

  int type;
  RC_CHECK_RETURN(dom->des(&type));
  m_elecType = static_cast<util::ElectricalType>(type);
  RC_CHECK_RETURN(dom->des(&type));
  m_pinType = static_cast<util::PinType>(type);

  RC_CHECK_RETURN(m_lineStyle.deserialize(dom));

  RC_CHECK_RETURN(m_pinNumText->deserialize(dom));
  RC_CHECK_RETURN(m_pinSymbolText->deserialize(dom));
  relayout();
  return 0;
}

} // namespace schematic

File: esim-frontend/test/schematic/elementPin.py
#
#  OpenESIM (Electrical Engineering Simulator)
#  Copyright (C) 2019, ESIM team. <diyer175@hotmail.com>
#
#  This project is free software; you can redistribute it and/or
#  modify it under the terms of the GNU Lesser General Public License(GPL)
#  as published by the Free Software Foundation; either version 2.1
#  of the License, or (at your option) any later version.
#
#  This project is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#  Lesser General Public License for more details.
#

from PyQt5.QtGui import QFontMetrics
from .element import Element, ElementSizeType, ElementDirection
from .styleTemplate import CreateLineStyle, CreateTextStyle
from util.persistence import save, saveObj, get, getObj
from util.ieeeTypes import ElectricalType, PinType

class ElementPin(Element):
    def __init__(self, doc, id=-1, x1=0, y1=0):
        super().__init__(doc, id)
        self.lineStyle = CreateLineStyle("pin")
        self.length = doc.GridX * 2
        self.symbolSpacing = 5
        self.numSpacing = 2
        self.symbol = "PIN"
        self.num = "1"
        self.angle = 0
        self.dir = ElementDirection.DIR_LEFT
        self.elecType = ElectricalType.ELEC_INPUT
        self.pinType = PinType.PIN_NONE
        self.x1 = x1
        self.y1 = y1
        self.lineX1 = self.lineY1 = self.lineX2 = self.lineY2 = 0
        self.pinSymbolText = None
        self.pinNumText = None
        
    # @Override
    def resolveSubElements(self, doc):
        if super().resolveSubElements(doc):
            # Deserialize from stream
            self.pinSymbolText = self.subElements[0]
            self.pinNumText = self.subElements[1]
        else:
            # Create text element for port symbol
            self.pinSymbolText = doc.createElement("text")
            self.pinSymbolText.setText(self.symbol)
            self.pinSymbolText.textStyle = CreateTextStyle("symbol")
            
            self.pinNumText = doc.createElement("text")
            self.pinNumText.setText(self.num)
            self.pinNumText.textStyle = CreateTextStyle("pin")
            
            self.attachElement(self.pinSymbolText)
            self.attachElement(self.pinNumText)

        self.setDirection(self.dir) # __updateLayout() as well
        
    def setPos(self, x1, y1):
        self.x1 = x1
        self.y1 = y1
        self.__updateLayout()
        
    def setDirection(self, dir):
        self.dir = dir
        if dir == ElementDirection.DIR_RIGHT:
            self.angle = 0
        elif dir == ElementDirection.DIR_TOP:
            self.angle = 90
        elif dir == ElementDirection.DIR_LEFT:
            self.angle = 180
        elif dir == ElementDirection.DIR_BOTTOM:
            self.angle = 270
        self.__updateLayout()
        
    def setSymbolSpacing(self, symbolSpacing):
        self.symbolSpacing = symbolSpacing
        self.__updateLayout()
    def setNumSpacing(self, numSpacing):
        self.numSpacing = numSpacing
        self.__updateLayout()
        
    def setLength(self, length):
        self.length = length
        self.__updateLayout()
        
    def setSymbol(self, symbol):
        self.symbol = symbol
        self.__updateLayout()
    def setNum(self, num):
        self.num = num
        self.__updateLayout()

    def setElecType(self, type):
        self.elecType = type
    def setPinType(self, type):
        self.pinType = type

    """
    Layout the pin paintings in four directions
    called when the layout has been changed.
    """
    def __updateLayout(self):
        self.pinSymbolText.setRotation(self.angle)
        self.pinNumText.setRotation(self.angle)
        
        self.lineX1, self.lineY1 = self.x1, self.y1
        
        # Layout:
        #            NumText(center)
        # [ ]  ( ) -----------------------   SymbolText
        #  |    |
        # note port symbol
        
        if self.dir == ElementDirection.DIR_LEFT:
            self.lineX2 = self.x1 - self.length
            self.lineY2 = self.y1
            
            cx = self.lineX1 - (self.length - self.pinNumText.textWidth)/2
            cy = self.lineY1 - self.pinNumText.textHeight - self.numSpacing
            self.pinNumText.setPos(cx, cy)
            
            cx = self.lineX2 - self.symbolSpacing
            cy = self.lineY2 - (self.pinNumText.textHeight/2)
            self.pinSymbolText.setPos(cx, cy)

        elif self.dir == ElementDirection.DIR_TOP:
            self.lineX2 = self.x1
            self.lineY2 = self.y1 + self.length
            
            cy = self.lineY1 + (self.length - self.pinNumText.textWidth)/2
            cx = self.lineX1 - self.numSpacing
            self.pinNumText.setPos(cx, cy)
            
            cy = self.lineY1 + self.length + self.symbolSpacing
            cx = self.lineX1 + (self.pinNumText.textHeight/2)
            self.pinSymbolText.setPos(cx, cy)

        elif self.dir == ElementDirection.DIR_RIGHT:
            self.lineX2 = self.x1 + self.length
            self.lineY2 = self.y1
            
            cx = self.lineX1 + (self.length - self.pinNumText.textWidth)/2
            cy = self.lineY1 - self.pinNumText.textHeight - self.numSpacing
            self.pinNumText.setPos(cx, cy)
            
            cx = self.lineX2 + self.symbolSpacing
            cy = self.lineY2 - (self.pinNumText.textHeight/2)
            self.pinSymbolText.setPos(cx, cy)
            
        elif self.dir == ElementDirection.DIR_BOTTOM:
            self.lineX2 = self.x1
            self.lineY2 = self.y1 - self.length
            
            cy = self.lineY2 + (self.length+self.pinNumText.textWidth)/2
            cx = self.lineX1 - self.pinNumText.textHeight - self.numSpacing
            self.pinNumText.setPos(cx, cy)
            
            cy = self.lineY2 - self.symbolSpacing
            cx = self.lineX2 - (self.pinNumText.textHeight/2)
            self.pinSymbolText.setPos(cx, cy)

        self._invalidBounding()
            
    # @Override
    def classname(self):
        return "pin"
        
    # @Override
    def serialize(self, node):
        super().serialize(node)
        save(node, [self.x1, self.y1, self.dir, self.symbol, self.num, self.elecType, self.pinType] )
        saveObj(node, self.lineStyle)
        
    # @Override
    def deserialize(self, node):
        super().deserialize(node)
        (self.x1, self.y1, self.dir, self.symbol, self.num, self.elecType, self.pinType) = get(node)
        getObj(node, self.lineStyle)
        
    # @Override
    def sizeType(self):
        return ElementSizeType.RECT_1P
        
    # @Override
    def bounding(self):
        return super().bounding()

    # @Override
    def render(self, p):
        lineContext = p.lineContext()
        p.setLineStyle(self.lineStyle)

        if self.modifiable():
            p.ellipse(self.lineX1 - 5, self.lineY1 - 5, 10, 10);
        p.line(self.lineX1, self.lineY1, self.lineX2, self.lineY2)
        
        p.setLineContext(lineContext)
        